rust/connlib/tunnel/src/lib.rs

```diff
@@ -415,6 +424,64 @@ where
     }
 }
 
+struct Io {
+    device: Device,
+    timeout: Option<Pin<Box<tokio::time::Sleep>>>,
+    sockets: Sockets,
+}
+
+impl Io {
+    fn poll<'b>(
+        &mut self,
+        cx: &mut Context<'_>,
+        buffer: &'b mut [u8],
+    ) -> Poll<io::Result<Input<'b, impl Iterator<Item = Received<'b>>>>> {
+        let (buf1, buf2) = buffer.split_at_mut(buffer.len() / 2); // If rustc borrow-checker would be better, we wouldn't need
+
+        if let Some(timeout) = self.timeout.as_mut() {
+            if timeout.poll_unpin(cx).is_ready() {
+                return Poll::Ready(Ok(Input::Timeout(timeout.deadline().into())));
+            }
+        }
+
+        if let Poll::Ready(network) = self.sockets.poll_recv_from(cx, buf1)? {
+            return Poll::Ready(Ok(Input::Network(network)));
+        }
+
+        if let Poll::Ready(packet) = self.device.poll_read(buf2, cx)? {
+            return Poll::Ready(Ok(Input::Device(packet)));
+        }
+
+        Poll::Pending
+    }
+
+    fn reset_timeout(&mut self, timeout: Instant) {
+        let timeout = tokio::time::Instant::from_std(timeout);
+
+        match self.timeout.as_mut() {
+            Some(existing_timeout) if existing_timeout.deadline() != timeout => {
+                existing_timeout.as_mut().reset(timeout)
+            }
+            Some(_) => {}
+            None => self.timeout = Some(Box::pin(tokio::time::sleep_until(timeout))),
+        }
+    }
+
+    fn send_network(&self, transmit: quinn_udp::Transmit) {
+        let _ = self.sockets.try_send_quinn(transmit);
+    }
+
+    fn send_device(&self, packet: IpPacket<'_>) {
+        let _ = self.device.write(packet);
+    }
+}
```

Comment:
@conectado I think using a generic `Io` component like this, we can eliminate most of the duplication between clients and gateways when it comes to the eventloop. What do you think?